src/slot.js

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  AmpContext,
+  withAmpContext,
+} from './amp-context.js';
+import {useMountEffect} from './amp-react-utils.js';
+
+const {
+  useContext,
+  useEffect,
+  useLayoutEffect,
+  useRef,
+} = preactHooks;
+
+
+/**
+ * @param {!Element} element
+ * @param {string} name
+ * @param {!Object|undefined} props
+ * @return {!ReactElement}
+ */
+export function createSlot(element, name, props) {
+  element.setAttribute('slot', name);
+  const slotProps = Object.assign({name}, props || {});
+  return preact.createElement(Slot, slotProps);
+}
+
+/**
+ * Slot component.
+ */
+export function Slot(props) {
```

Comment:
It's really difficult to tell what's changed because of the file move.

Reply:
Good point. I created a preliminary PR with just a plain factoring out of Slot component. Please review: https://github.com/jridgewell/amp-react-prototype/pull/31. I'll rebase this PR against that one.